From: Rainer Weikusat <rweikusat@mssgmbh.com>
Date: Mon, 6 Dec 2010 16:33:43 +0100
Subject: [PATCH] restore lxc.mount/lxc.mount.entry functionality

The lxc_setup-routine in conf.c tries to configure the filesystem
namespace for a new container by first recursively bind-mounting the
directory which is supposed to become the root filesystem of the
container to the 'rootfs mount point' and then adding whatever other
mounts have been requested in the configuration file being used to the
original filesystem location of the nascent container root
filesystem. At least for Linux 2.6.35.7 (and .4), this doesn't work
because the MS_REC flag only affects submounts which already existed
at the time of the bind mount, not new ones which come into existence
afterwards. The patch moves the setup_rootfs call in lxc_setup behind
the two 'setup additional mounts' calls which fixes this problem (at
least for me).

Signed-off-by: Rainer Weikusat <rweikusat@mssgmbh.com>
Signed-off-by: Guido Trotter <ultrotter@debian.org>
---
 src/lxc/conf.c |   10 +++++-----
 1 files changed, 5 insertions(+), 5 deletions(-)

diff --git a/src/lxc/conf.c b/src/lxc/conf.c
index e4e4bb5..20a843c 100644
--- a/src/lxc/conf.c
+++ b/src/lxc/conf.c
@@ -1597,11 +1597,6 @@ int lxc_setup(const char *name, struct lxc_conf *lxc_conf)
 		return -1;
 	}
 
-	if (setup_rootfs(&lxc_conf->rootfs)) {
-		ERROR("failed to setup rootfs for '%s'", name);
-		return -1;
-	}
-
 	if (setup_mount(lxc_conf->fstab)) {
 		ERROR("failed to setup the mounts for '%s'", name);
 		return -1;
@@ -1612,6 +1607,11 @@ int lxc_setup(const char *name, struct lxc_conf *lxc_conf)
 		return -1;
 	}
 
+	if (setup_rootfs(&lxc_conf->rootfs)) {
+		ERROR("failed to setup rootfs for '%s'", name);
+		return -1;
+	}
+
 	if (setup_cgroup(name, &lxc_conf->cgroup)) {
 		ERROR("failed to setup the cgroups for '%s'", name);
 		return -1;
-- 
